api/src/main/java/org/pmiops/workbench/cohortreview/AnnotationQueryBuilder.java

```diff
@@ -0,0 +1,222 @@
+        orderByBuilder.append(orderByColumn);
+      } else {
+        String columnAlias = columnAliasMap.get(orderByColumn);
+        if (columnAlias == null) {
+          throw new BadRequestException("Column " + orderByColumn + " in orderBy must be present in columns");
+        }
+        orderByBuilder.append(columnAlias);
+      }
+    }
+    return orderByBuilder.toString();
+  }
+
+  private String getLimitAndOffsetSql(int limit, long offset) {
+    StringBuilder endSqlBuilder = new StringBuilder("\nLIMIT ");
+    endSqlBuilder.append(limit);
+    if (offset != 0L) {
+      // TODO: consider pagination based on values rather than offsets
+      endSqlBuilder.append(" OFFSET ");
+      endSqlBuilder.append(offset);
+    }
+    return endSqlBuilder.toString();
+  }
+
+  private String getSql(CohortReview cohortReview, List<CohortStatus> statusFilter,
+      AnnotationQuery annotationQuery, int limit, long offset,
+      Map<String, CohortAnnotationDefinition> annotationDefinitions) {
+    Map<String, String> columnAliasMap = Maps.newHashMap();
+    String selectAndFromSql = getSelectAndFromSql(cohortReview, annotationQuery.getColumns(),
+        annotationDefinitions, columnAliasMap);
+    String whereSql = getWhereSql(cohortReview, statusFilter);
+    String orderBySql = getOrderBySql(annotationQuery.getOrderBy(), columnAliasMap);
+    String limitAndOffsetSql = getLimitAndOffsetSql(limit, offset);
+        StringBuilder sqlBuilder = new StringBuilder(selectAndFromSql);
+    sqlBuilder.append(whereSql);
+    sqlBuilder.append(orderBySql);
+    sqlBuilder.append(limitAndOffsetSql);
+    return sqlBuilder.toString();
+  }
+
+  public Iterable<Map<String, Object>> materializeAnnotationQuery(CohortReview cohortReview,
+      List<CohortStatus> statusFilter,
+      AnnotationQuery annotationQuery, int limit, long offset) {
+    Map<String, CohortAnnotationDefinition> annotationDefinitions = null;
+    List<String> columns = annotationQuery.getColumns();
+    if (columns == null || columns.isEmpty()) {
+      // By default get person_id, review_status, and all the annotation definitions.
+      columns = new ArrayList<>();
+      columns.add(PERSON_ID_COLUMN);
+      columns.add(REVIEW_STATUS_COLUMN);
+      annotationDefinitions = getAnnotationDefinitions(cohortReview);
+      columns.addAll(annotationDefinitions.values().stream()
+          .map(CohortAnnotationDefinition::getColumnName).collect(Collectors.toList()));
+      annotationQuery.setColumns(columns);
+    }
+    List<String> orderBy = annotationQuery.getOrderBy();
+    if (orderBy == null || orderBy.isEmpty()) {
+      annotationQuery.setOrderBy(ImmutableList.of(PERSON_ID_COLUMN));
+    }
+    String sql = getSql(cohortReview, statusFilter, annotationQuery, limit, offset,
+        annotationDefinitions);
+    return namedParameterJdbcTemplate.query(sql, new RowMapper<Map<String, Object>>() {
```

Comment:
Seems that we are using a namedParameterJdbcTemplate but were not passing any named parameters? Should we set params using notation :myNamedParam instead of String.format?

Reply:
Done.

Note we still using String.format() for alias name substitution, etc. but now the integer values that were being inlined before are bound parameters.